### PR TITLE
Softmax MKLDNN FLUID operator

### PR DESCRIPTION
**Pull request by jczaja**
This PR provides MKLDNN based Softmax op implementation. 

### Performance and testing:
On tested models softmax MKLDNN op is roughly ~10x faster than Plain CPU version.
RNN Search (https://github.com/dzhwinter/benchmark/blob/master/fluid/machine_translation.py)
executes training in ~90% of Plain CPU time. RNN search does converge with Softmax MKLDNN op used.

### Notes
- It was needed to update cross_entropy grad op with code preventing -INF results in a similar way as it was done in cross_entropy op.

### CODE REVIEW

**Comment by luotao1** [comment]
LGTM. @tensor-tang Could you help review the `softmax_mkldnn_op.cc`?

File: paddle/fluid/operators/softmax_mkldnn_op.cc [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+using paddle::platform::MKLDNNMemDesc;
+
+using mkldnn::memory;  // Note: paddle has also "memory" namespace
+using mkldnn::primitive;
+using mkldnn::softmax_forward;
+using mkldnn::prop_kind;
+using mkldnn::stream;
+
+template <typename T>
+class SoftmaxMkldnnKernel : public paddle::framework::OpKernel<T> {
```
Comment (tensor-tang) [inline paddle/fluid/operators/softmax_mkldnn_op.cc]: `SoftmaxMkldnnKernel` ==> `SoftmaxMKLDNNKernel`

File: python/paddle/fluid/layers/nn.py [hunk 2]
```diff
@@ -162,8 +163,11 @@ def fc(input,
             inputs={"X": input_var,
                     "Y": w},
             outputs={"Out": tmp},
-            attrs={"x_num_col_dims": num_flatten_dims,
-                   "y_num_col_dims": 1})
+            attrs={
+                "x_num_col_dims": num_flatten_dims,
+                "y_num_col_dims": 1,
+                'use_mkldnn': use_mkldnn
```
Comment (tensor-tang) [inline python/paddle/fluid/layers/nn.py]: Why softmax PR would change fc?

File: python/paddle/fluid/layers/nn.py [hunk 3]
```diff
@@ -81,6 +81,7 @@ def fc(input,
        num_flatten_dims=1,
        param_attr=None,
        bias_attr=None,
+       use_mkldnn=False,
```
Comment (tensor-tang) [inline python/paddle/fluid/layers/nn.py]: Why softmax PR would change fc?

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (tensor-tang) [inline python/paddle/fluid/layers/nn.py]: Save as above, this should be added in FC PR, right?

Comment (jczaja) [reply python/paddle/fluid/layers/nn.py]: @tensor-tang You are correct, but without those changes to FC Softmax MKLDNN would remain unused code. Till FC of MKLDNN is not ready (it is under development) then there is no other option for users to take advantage of softmax mkldnn operator  as it is usually used as activation of FC layer. Do you suggest to remove use_mkldnn attribute/param from FC ?

File: paddle/fluid/operators/softmax_mkldnn_op.cc (hunk 1 above)
Comment (jczaja) [reply paddle/fluid/operators/softmax_mkldnn_op.cc]: Ok

File: python/paddle/fluid/layers/nn.py (hunk 2 above)
Comment (tensor-tang) [reply python/paddle/fluid/layers/nn.py]: I think there should be some other ways to use mkldnn softmax without change fc. But since we will implement MKLDNN FC soon, I think it's fine, you do not need revert back now.
